### ISSUE TITLE
Select Box Link not populating from external xml

### ISSUE THREAD

**Issue by jude86-7** [symphonists/xmlimporter#50]
I have two sections; 

one titled 'category' with a text input field for the title.
another titled 'items' with a text input field for the title and a select box link for the category (not multiple option).

My xml looks like this;

``` xml
<?xml version="1.0" encoding="UTF-8"?>
<test_element>
    <entry>
        <title>Selected One</title>
        <category id="1105">Item One</category>
    </entry>
    <entry>
        <title>Selected Two</title>
        <category id="1105">Item One</category>
    </entry>
    <entry>
        <title>Selected Three</title>
        <category id="1106">Item Two</category>
    </entry>
</test_element>
```

The xml importer uses;

included elements: /test_element/entry
fields: title/text() and category/@id

When run this brings the titles in but doesn't bring in the category. 

I am running symphony 2.6.1 and xml importer version 2.2

Is this a compatibility issue or am I missing something in the process?

Many thanks in advance

**Editied to support code block**

**Comment by nitriques** [comment]
I've never tried this, sadly. @jonmifsud would you have an idea ?

**Comment by jude86-7** [comment]
Bump, has anyone else come across this problem? I'm struggling to find an answer to this. I've created the exact same set up on an older installation of symphony and it works as expected. So it seems to be an issue in the later versions in my case 2.6.1.

**Comment by nitriques** [comment]
> on an older installation of symphony

With the same version of the extension ? Which old version ?

**Comment by jude86-7** [comment]
ah no, my bad sorry. the extension version was 2.1.2 and symphony 2.3.6.
I could perhaps install and test on symphony 2.5.0 and xml importer 2.2 if that would help?

**Comment by nitriques** [comment]
Yes please: that way, we would know which repo contains the regression

**Comment by jude86-7** [comment]
okay no problem, I'll give it a go and post back any findings. thanks for your help

**Comment by nitriques** [comment]
:+1:

**Comment by jude86-7** [comment]
sorry deadlines have kept me away from this over the last couple of days. I've set up the following clean installations and got the following results;

symphony 2.5
xml importer 2.2
working as expected - select box link values go into correct entries - all okay

symphony 2.6.1
xml importer 2.2
a little further with this - on a blank install the xml importer imports the entries with values for the select box link (on earlier testing they were blank) however no matter what is in the xml file the select box link value is always the first option.

symphony 2.6.7
xml importer 2.2
behaves in the same way as 2.6.1 - select box link has a value but it's always the first entry regardless of what's in the xml file.

I hope this helps narrow things down a little!

**Comment by nitriques** [comment]
Yes thanks!

**Comment by twiro** [comment]
Populating SBL with entry-IDs via XML-Importer defintively used to work in the 2.3.x-days - I posted [my test results](http://www.getsymphony.com/discuss/thread/45612/#position-8) in the Symphony Forum back then.

Maybe the discussion over there is helpful for you - I haven't used the importer for quite a while and won't find the time to repeat my tests with a current setup at the moment.

And while thinking about it...

> _Is this a **compatibility issue** or am I missing something in the process?_

... I guess you might be right - have you tried using the [integration-branch](https://github.com/symphonists/xmlimporter/tree/integration) in Symphony 2.6.x? Given the fact that it's **40 commits ahead of master** I suppose most of the stuff that has been done to make XML Importer compatible with the latest Symphony releases hasn't found it's way into an official release yet.

@nitriques , @brendo , @jonmifsud - anyone able to change that and take care of releasing a new version? Would be about time after 2 years :) I'm sorry I can't help, but I'm just getting started with extension development and this is totally over my head (for now)...

**Comment by jonmifsud** [comment]
@twiro / @jude86-7 this one's on me. Integration branch is def what has to be used, it uses a datasource to simplify the process. I've used it for over a year now without issues, had a very large migration around this time last year where I did a bunch of updates. I'm swamped at the moment but I'll try to allocate some time to make sure it's released. 

@jude86-7 I've most certainly used the association field (which has pretty much the same codebase as SBL) and using entry-ids should work.

**Comment by twiro** [comment]
> _...but I'll try to allocate some time to make sure it's released._

That's awesome - thank you very much Jon!
I guess a new release would also clean up the issue list a little bit :)

**Comment by jude86-7** [comment]
thanks very much for your help. I did try the integration branch but couldn't get it to work, I can't remember the specific issue but I'll give it another go on a clean install and see how I get on. Thanks!

**Comment by jude86-7** [comment]
thanks for your help and suggestions. I've created a blank install of symphony 2.6.7 and installed the integration branch of xml importer 3.0.

All seems to be working okay, except for the select box link. Using ids in the xml the select box link field is still blank when I import the items.

@jonmifsud just wondering if I am missing something here - which version of symphony worked with the association field and entry ids for you?

**Comment by jean-luc** [comment]
When `$field->prepareImportValue()` is called on a selectbox_link field it's setting the wrong `relation_id` in the database. In my test the `relation_id` value was always `1` when it should be ID of the linked entry.

I have got the importer working by changing the following line in the integration branch version to prevent `prepareImportValue()` from being called for SBL:
https://github.com/symphonists/xmlimporter/blob/integration/lib/class.xmlimporter.php#L283

```
if ($field->get('type') != 'selectbox_link') {
    $value = $field->prepareImportValue($value, $mode, $entry->get('id'));
}
```

With the above change the Select Box Link fields imports correctly. I hope is helpful in finding the actual issue and someone can patch in a proper fix.

**Comment by nitriques** [comment]
@jean-luc thanks for this. Can you try to see if it's `selectbox_link`'s `prepareImportValue` that is wrong ?

**Comment by jude86-7** [comment]
@jean-luc thank you this helped me a lot. I made the code change and can confirm the select box link is working for me now on symphony 2.6.7 with xml importer 3.0. thanks to everyone for their help. I'll keep checking back to see if a patch is applied for the future.

**Comment by nitriques** [comment]
@jude86-7 @jean-luc I think I've found the problem. The xmlimporter relies on the fact that the `getImportModes()` implementation of the `selectbox_link` should have the STRING_VALUE before the ARRAY_VALUE... but @brendo did changed it to be the exact opposite: https://github.com/symphonycms/selectbox_link_field/commit/4bdd68cc477dcdaee84c14ee7c5a3f25004013c1

Should the importer specify which export mode to use instead of randomly picking the first one ?

**Comment by cylkee** [comment]
I think I'm suffering from this issue too.  Bit gutted I've spent the best part of a hot day trawling the forum/Gitter and trying different things/XMLImporterHelpers etc.

I too am using Symphony 2.6.7 with XML Importer 2.2.  Prior to this I've done a successful importation of some other data, so I reckon I've got a good understanding of what I need to input.  Here's fundamentally what's failing:

Section A:
- ID
- Title

Section B:
- SBL [Section A: Title]

I was originally trying to import data which doesn't know Symphony IDs, but I generated my own XML from Symphony which connects the title with ID just so I could test:

`<?xml version="1.0" encoding="UTF-8"?>
<posts>
  <post>
    <id>337</id>
    <title>This is a title</title>
  </post>
  <post>
    <id>336</id>
    <title>This is a another title</title>
  </post>
</posts>`

Import to Section B:
- SBL
  - XPath Expression: title/text()
  - PHP Function: 
  - Is Unique: yes

I tried id/text() too.  Also, if I put into the XPath Expression field for the SBL a known plain-text Title, handle or id, it does not make the link.

**Comment by jude86-7** [comment]
@cylkee have you tried using the integration branch 3.0 of the xml importer? This was the one I needed to use to get it working, along with the fix Jean-Luc provides above.

With regard to setting up the xml importer my xml looked like this for the sbl link item;

`<Category id="13" handle="title">title</Category>`

So in the destination fields my xpath expression read;

`Category/@id`

I hope this helps

**Comment by nitriques** [comment]
@jude86-7 Can you send a PR ? I am glad you were able to fix it !

**Comment by cylkee** [comment]
@jude86-7 No, not tried IB. It's not super urgent for me, so I'd rather wait until a proper release to save faffing.

I could try that XML, but pretty sure my XML/XPath would work too.  Thanks.

**Comment by jude86-7** [comment]
@nitriques I can't take the credit I'm afraid. I just used the code suggested by @jean-luc and it worked for me. As far as I'm aware though this was just a temporary fix or a patch as Jean-Luc states;

 _'I hope is helpful in finding the actual issue and someone can patch in a proper fix.'_

So I'm not sure if a PR would be the right thing here? Saying that I'm fairly new to all of this!

**Comment by nitriques** [comment]
@jude86-7 Yes try to send a PR! I'll help you if you need help, no worries !! Please start from (and push) against the `integration` branch ;) I am on gitter from 10 to 17 EDT each day.

**Comment by jude86-7** [comment]
@nitriques Sorry for not coming back to you sooner, I've had a few deadlines to finish up.
I've created a PR, hope I've done it right! Just let me know if I need to do anything different!
Thanks

**Comment by nitriques** [comment]
> Sorry for not coming back to you sooner, I've had a few deadlines to finish up.

Don't be sorry, that's perfectly ok ! I'll check you PR asap.

**Comment by nitriques** [comment]
@jude86-7 I've looked into things and the current code base should work. Which version of the SBL extension are you running ?

**Comment by nitriques** [comment]
And before we do anything, can you add a `var_dump($mode, $value);die;` before [this line](https://github.com/symphonists/xmlimporter/pull/52/files#diff-1bbcdec62ba2d01ecbf2a2c6d13847adR284)

**Comment by jude86-7** [comment]
SBL extension is 1.34, XML importer extension 3.0.0 and Symphony version 2.6.1
I've added in the line too, hope this helps.
Thanks

**Comment by nitriques** [comment]
> I've added in the line too, hope this helps.

The line was only for debug purpose, not to be committed. What's the output you get ?

Also, I think you do not need the fix if you update SBL to 2.0.0.

**Comment by jude86-7** [comment]
> The line was only for debug purpose, not to be committed. What's the output you get ?

Oops sorry my bad, I'm still new at all of this but learning a lot!

That's great thank you, I'm a bit short on time at the moment but I'll definitely give SBL 2.0.0 a go without the fix to the importer.

**Comment by nitriques** [comment]
This should be now fixed as per #52
